### PR TITLE
App Data Retriever and Search Term Explorer Merge into Develop

### PR DESCRIPTION
**Pull request by sauyon**

### CODE REVIEW

**Comment by AdamSlack** [comment]
got 2 and a half issues to resolve first :haircut_man:

File: pipeline/scraper/db.js [hunk 1]
```diff
@@ -45,6 +45,44 @@ async function insertDev(dev) {
 }
 
 module.exports = {
+    /**
+     *  Query the search_terms table to get a list of terms that are stale
+     */
+    get_search_terms: async() => {
+        logger.debug('Fetching Search Terms');
+        var res = await query('SELECT search_term FROM search_terms WHERE current_date > last_searched + interval \'1 month\'');
```
Comment (sauyon) [inline pipeline/scraper/db.js]: Might use `...WHERE age(last_searched) > interval \'1 month\'` instead.

File: pipeline/scraper/db.js [hunk 2]
```diff
@@ -45,6 +45,44 @@ async function insertDev(dev) {
 }
 
 module.exports = {
+    /**
+     *  Query the search_terms table to get a list of terms that are stale
+     */
+    get_search_terms: async() => {
```
Comment (sauyon) [inline pipeline/scraper/db.js]: Please rename to `getSearchTerms`

**Review by sauyon — COMMENTED** [review]
I think I solved one of them :D

**Comment by sauyon** [comment]
But yeah I'll look at this later then.

**Comment by AdamSlack** [comment]
Sounds good! That should resolve the issue with current_date not being liked anymore for postgres

**Comment by AdamSlack** [comment]
Also, will go through and camel case over underscore all var names. Dunno why I swapped, just force of habit I guess :L

File: pipeline/scraper/db.js (hunk 2 above)
Comment (sauyon) [reply pipeline/scraper/db.js]: @AdamSlack would prefer you reply in this thread :rofl: anyway yeah snake casing is the best :+1: but everybody seems to like camel casing here for some reason.

**Comment by AdamSlack** [comment]
Should be ready to merge now, pls review! :+1: :+1:

File: pipeline/scraper/db.js [hunk 3]
```diff
@@ -31,20 +31,92 @@ function connect() {
 }
 
 async function insertDev(dev) {
-    var res = await query('SELECT id FROM developers WHERE $1 = ANY(email)', [dev.email]);
-    if (res.rowCount > 0) {
-        return res.rows[0].id;
-    }
+    try {
```
Comment (sauyon) [inline pipeline/scraper/db.js]: remove try catch

File: pipeline/scraper/db.js [hunk 4]
```diff
@@ -31,20 +31,92 @@ function connect() {
...
+    } catch (err) { logger.err(err); }
     return res.rows[0].id;
 }
 
+// TODO: change _ names to camelCase names.
 module.exports = {
+    /**
+     *  Query the search_terms table to get a list of terms that are stale
+     */
+    //TODO: Rename to 'getStaleSearchTerms'
```
Comment (sauyon) [inline pipeline/scraper/db.js]: remove TODO lines

File: pipeline/scraper/retriever.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+/**
+ *  Example of promise to fetch permissions from Google Play Store.
+ */
+// gplay.permissions({
+//     appId: 'com.dxco.pandavszombies',
+//     short: true
+// }).then(
+//     (app) => logger.info(app),
+//     (err) => logger.err(err.message)
+// );
```
Comment (sauyon) [inline pipeline/scraper/retriever.js:73]: Trailing newlines please.

File: pipeline/scraper/scrape.js [hunk 6]
```diff
@@ -62,7 +62,8 @@ let waitForPipe = () => {
     } else {
         main();
     }
-}; waitForPipe();
+};
+waitForPipe();
```
Comment (sauyon) [inline pipeline/scraper/scrape.js]: I would move this as close to the end as possible

File: pipeline/scraper/logger.js [hunk 7]
```diff
@@ -6,18 +6,18 @@ Simply logger to correctly error messages to systemd.
...
 const util = require('util');
 const EMERG = 0,
-      ALERT = 1,
-      CRIT = 2,
-      ERR = 3,
-      WARNING = 4,
-      NOTICE = 5,
-      INFO = 6,
-      DEBUG = 7;
+    ALERT = 1,
```
Comment (sauyon) [inline pipeline/scraper/logger.js:9]: should be aligned with const

**Review by sauyon — COMMENTED** [review]
dammit guys

File: pipeline/example_config.json [hunk 8]
```diff
@@ -1,25 +1,38 @@
...
+        "db": {
+            "user": "downloader",
+            "password": ""
+        }
+
+    },
+    "analyzer": {
+        "db": {
+            "user": "analyzer",
+            "password": ";:&!kQu{o>=F+Mm)lnkTaL{_U@&<$I"
```
Comment (sauyon) [inline pipeline/example_config.json]: goddammit guys

**Review by sauyon — COMMENTED** [review]
...

File: db/init_db.sql [hunk 9]
```diff
@@ -30,19 +30,25 @@ create table playstore_apps(
...
   video                  text                                                 ,
   recent_changes       text[]                                                 ,
-  crawl_date             date                                         not null
+  crawl_date             date                                         not null,
+  permissions          text[]
+);
+
+create table search_terms(
+  search_term            text                              primary key not null,
+  last_searched          DATE                                          not null
```
Comment (sauyon) [inline db/init_db.sql:52]: why is DATE the only thing capitalized in the entire file lol

File: pipeline/scraper/explorer.js [hunk 10]
```diff
@@ -4,12 +4,24 @@
...
-var fs = require('fs-extra');
-var logger = require('./logger.js');
-var wordStoreLocation = config.wordStashDir + '/suggested_words.txt';
+const alphabet = require('alphabet');
+const _ = require('lodash');
+const config = require('/etc/xray/config.json');
+const fs = require('fs-extra');
+const logger = require('./logger.js');
+const wordStoreLocation = config.datadir + '/suggested_words.txt';
+var database = require('./db.js');
```
Comment (sauyon) [inline pipeline/scraper/explorer.js]: why is this the only thing that's var

File: pipeline/scraper/explorer.js [hunk 11]
```diff
@@ -36,6 +48,27 @@ function writeScrapedWords(word, location) {
     });
 }
 
+
+/**
+ * Used returns an array where each line is a search term.
+ * @param {*the location of the file that is to be read} file_location 
+ */
+function openSearchTerms(file_location) {
```
Comment (sauyon) [inline pipeline/scraper/explorer.js:48]: are these two functions necessary now?

File: pipeline/scraper/explorer.js [hunk 12]
```diff
@@ -59,27 +92,29 @@ function cartesianProductChars() {
  * 
  * @param {*The list of words used to get autocompletes} startingWords 
  */
+// TODO: Store scraped word to the Database not txt
 function scrapeSuggestedWords(startingWords) {
```
Comment (sauyon) [inline pipeline/scraper/explorer.js:88]: I think this should be slightly smarter. I'll open an issue for it.

File: pipeline/scraper/logger.js [hunk 13]
```diff
@@ -45,4 +45,4 @@ module.exports = {
     debug: function() {
         log(DEBUG, arguments);
     }
-};
+};
```
Comment (sauyon) [inline pipeline/scraper/logger.js:49]: trailing newlines!

File: pipeline/scraper/scrape.js [hunk 14]
```diff
@@ -345,4 +344,4 @@ function main() {
 //         let p = require('path');
 //         fs.createReadStream(p.join(wordStash,file)).pipe(parser);
 //     });
-// });
+// });
```
Comment (sauyon) [inline pipeline/scraper/scrape.js]: trailing newlines!

File: pipeline/scraper/retriever.js [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+var db = new database('explorer');
+
+const region = 'us';
+
+/**
+ * Inserts app data into the db using db.js
+ * @param {*The app data json that is to be inserted into the databae.} app_data 
+ */
+function insertAppData(app_data) {
+
```
Comment (sauyon) [inline pipeline/scraper/retriever.js:17]: get rid of this newline.

File: pipeline/scraper/retriever.js [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+    //Checking version data - correct version to update date
+    if (!app_data.version || app_data.version === 'Varies with device') {
```
Comment (sauyon) [inline pipeline/scraper/retriever.js:19]: maybe we should start tagging all versions with updated date instead?

File: pipeline/scraper/retriever.js [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+        logger.debug('Version not found defaulting too', app_data.updated);
+        let formatDate = app_data.updated.replace(/\s+/g, '').replace(',', '/');
+        app_data.version = formatDate;
+    }
+
+    // push the app data to the DB
+    return db.insertPlayApp(app_data, region);
+}
+
+function updateSearchedTermDate(searchTerm) {
```
Comment (sauyon) [inline pipeline/scraper/retriever.js]: this really doesn't need to be a function - that log should probably go into the db module.

File: pipeline/scraper/retriever.js (hunk 16 above)
Comment (sauyon) [inline pipeline/scraper/retriever.js:19]: Also I think this should probably go into the db module.

File: pipeline/scraper/retriever.js [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+        } else {
+            logger.debug('App already existing', app_data.appId);
+        }
+    }));
+}
+
+/**
+ * uses db.js to fetch search terms from the database.
+ */
+async function fetch_search_terms() {
```
Comment (sauyon) [inline pipeline/scraper/retriever.js]: ...why is this even a function

**Review by sauyon — COMMENTED** [review]
What's with the `json_dumps` directory?

Also we should nuke a lot of massive commented out sections that we don't need anymore.

File: pipeline/scraper/logger.js (hunk 7 above)
Comment (AdamSlack) [reply pipeline/scraper/logger.js:9]: blame the linter.

File: pipeline/scraper/retriever.js (hunk 5 above)
Comment (AdamSlack) [reply pipeline/scraper/retriever.js:73]: Added EOF new lines

File: db/init_db.sql (hunk 9 above)
Comment (AdamSlack) [reply db/init_db.sql:52]: lol jk fixed.

File: pipeline/scraper/explorer.js (hunk 11 above)
Comment (AdamSlack) [reply pipeline/scraper/explorer.js:48]: not in use right now, but if we wanted to add external files of search terms. really they should be in a separate file as exports i guess!

File: pipeline/scraper/logger.js (hunk 13 above)
Comment (AdamSlack) [reply pipeline/scraper/logger.js:49]: blame VS Code. gotta run through linter with --fix. fml :disappointed:

File: pipeline/scraper/retriever.js (hunk 15 above)
Comment (AdamSlack) [reply pipeline/scraper/retriever.js:17]: Yes Sir. Done Sir. :vulcan_salute:

File: pipeline/scraper/retriever.js (hunk 16 above)
Comment (AdamSlack) [reply pipeline/scraper/retriever.js:19]: Might be wise. Perhaps should be revisited in the future.

Comment (AdamSlack) [reply pipeline/scraper/retriever.js:19]: Agreed. Should revisit later after Ret' Exp' and Down' are all up and running in an acceptable state.

**Comment by AdamSlack** [comment]
Responded to all Comments, pls review :bowing_man:

**Review by AdamSlack — APPROVED** [review]
gone through and made the proposed changes

**Pull request merged**